sys/include/net/nanocoap.h

```diff
@@ -951,6 +950,18 @@ static inline unsigned coap_get_id(coap_pkt_t *pkt)
     return ntohs(pkt->hdr->id);
 }
 
+/**
+ * @brief   Get the start of data after the header
+ *
+ * @param[in]   hdr   Header of CoAP packet in continuous memory
+ *
+ * @returns     pointer to first byte after the header
+ */
+static inline uint8_t *coap_data_ptr(coap_hdr_t *hdr)
+{
+    return ((uint8_t *)hdr) + sizeof(coap_hdr_t);
```

Comment:
I'm not sure that this always results in the same pointer being returned as when using the flexible array member.

Reply:
Why is that? alignment?

Is this better?
```suggestion
    return (uint8_t *)(&hdr[1]);
```

Reply:
I wrote a minimal test, this indeed does result in the same address being returned, because of the packed attribute and size of the struct. I guess we won't change the coap hdr struct that often...

Reply:
here's my testcase:

```
[kaspar@ng tmp]$ gcc t.c -o t && ./t                                                                                                         
coap_data_ptr: 0x6
[kaspar@ng tmp]$ vim xc^C
[kaspar@ng tmp]$
[kaspar@ng tmp]$
[kaspar@ng tmp]$ vim t.c
[kaspar@ng tmp]$ cat t.c
#include <stdio.h>
#include <stdint.h>

typedef struct __attribute__((packed)) {
    uint8_t ver_t_tkl;          /**< version, token, token length           */
    uint8_t code;               /**< CoAP code (e.g.m 205)                  */
    uint16_t id;                /**< Req/resp ID                            */
#ifdef FLEXMEMB
    uint8_t data[];             /**< convenience pointer to payload start   */
#endif
} coap_hdr_t;

static inline uint8_t *coap_data_ptr(coap_hdr_t *hdr)
{
    return ((uint8_t *)hdr) + sizeof(coap_hdr_t);
}

int main() {
    coap_hdr_t hdr;
#ifdef FLEXMEMB
    printf("flex memb: %p\n", ((void*)&hdr.data) - (void*)&hdr);
#else
    printf("coap_data_ptr: %p\n", (void*)coap_data_ptr(&hdr) - (void*)&hdr);
#endif
    return 0;
}
[kaspar@ng tmp]$ gcc -DFLEXMEMB t.c -o t && ./t
flex memb: 0x4
[kaspar@ng tmp]$ gcc t.c -o t && ./t
coap_data_ptr: 0x4
[kaspar@ng tmp]$
```

Here's the fail case I was talking about:

```
[kaspar@ng tmp]$ cat t.c
#include <stdio.h>
#include <stdint.h>

typedef struct {
    uint8_t ver_t_tkl;          /**< version, token, token length           */
    uint8_t code;               /**< CoAP code (e.g.m 205)                  */
    uint16_t id;                /**< Req/resp ID                            */
    uint8_t a;
#ifdef FLEXMEMB
    uint8_t data[];             /**< convenience pointer to payload start   */
#endif
} coap_hdr_t;

static inline uint8_t *coap_data_ptr(coap_hdr_t *hdr)
{
    return ((uint8_t *)hdr) + sizeof(coap_hdr_t);
}

int main() {
    coap_hdr_t hdr;
#ifdef FLEXMEMB
    printf("flex memb: %p\n", ((void*)&hdr.data) - (void*)&hdr);
#else
    printf("coap_data_ptr: %p\n", (void*)coap_data_ptr(&hdr) - (void*)&hdr);
#endif
    return 0;
}
[kaspar@ng tmp]$ gcc -DFLEXMEMB t.c -o t && ./t
flex memb: 0x5
[kaspar@ng tmp]$ gcc t.c -o t && ./t
coap_data_ptr: 0x6
[kaspar@ng tmp]$
```

Reply:
The second case (structure not packed) can be handled by defining a second struct like:
```c
typedef struct  {coap_hdr_t hdr; uint8_t data0} coap_hdr_aux_t;
```
And then casting `coap_hdr_t*` to `coap_hdr_aux_t*` and taking the address of `data0`. Of course this also works for the packed struct.